packages/bundle-source/src/tool.js

```diff
@@ -0,0 +1,204 @@
+    readText: () => fs.promises.readFile(fileName, 'utf-8'),
+    neighbor: ref => make(path.resolve(fileName, ref)),
+    stat: () => fs.promises.stat(fileName),
+    absolute: () => path.normalize(fileName),
+    relative: there => path.relative(fileName, there),
+  });
+};
+
+export const makeFileWriter = (fileName, { fs, path }) => {
+  const make = there => makeFileWriter(there, { fs, path });
+  return harden({
+    toString: () => fileName,
+    writeText: txt => fs.promises.writeFile(fileName, txt),
+    readOnly: () => makeFileReader(fileName, { fs, path }),
+    neighbor: ref => make(path.resolve(fileName, ref)),
+    mkdir: opts => fs.promises.mkdir(fileName, opts),
+  });
+};
+
+export const exploreBundle = async endoZipBase64 => {
+  const zip = new ZipReader(new Uint8Array(decodeBase64(endoZipBase64)));
+  for await (const [name, _entry] of zip.files.entries()) {
+    console.log({ name });
+  }
+};
+
+export const makeBundleCache = (wr, cwd, readPowers) => {
+  const add = async (rootPath, targetName) => {
+    console.log(`${wr}`, 'add:', targetName, 'from', rootPath);
+    const srcRd = cwd.neighbor(rootPath);
+
+    const modTimeByPath = new Map();
+
+    const loggedRead = async loc => {
+      if (!loc.match(/\bpackage.json$/)) {
+        try {
+          const itemRd = cwd.neighbor(new URL(loc).pathname);
+          const ref = srcRd.relative(itemRd.absolute());
+          const { mtime } = await itemRd.stat();
+          modTimeByPath.set(ref, mtime);
+          // console.log({ loc, mtime, ref });
+        } catch (oops) {
+          console.error(oops);
+        }
+      }
+      return readPowers.read(loc);
+    };
+    const bundle = await bundleSource(
+      rootPath,
+      {},
+      { ...readPowers, read: loggedRead },
+    );
+
+    const { moduleFormat } = bundle;
+    assert.equal(moduleFormat, 'endoZipBase64');
+
+    const code = `export default ${JSON.stringify(bundle)};`;
+    await wr.mkdir({ recursive: true });
+    const bundleFileName = `bundle-${targetName}.js`;
+    const bundleWr = wr.neighbor(bundleFileName);
+    await bundleWr.writeText(code);
+    const { mtime: bundleTime } = await bundleWr.readOnly().stat();
+
+    const meta = {
+      bundleFileName,
+      bundleTime: bundleTime.toISOString(),
+      moduleSource: {
+        relative: bundleWr.readOnly().relative(srcRd.absolute()),
+        absolute: srcRd.absolute(),
+      },
+      contents: [...modTimeByPath.entries()].map(([relativePath, mtime]) => ({
+        relativePath,
+        mtime: mtime.toISOString(),
+      })),
+    };
+
+    await wr
+      .neighbor(`bundle-${targetName}-meta.json`)
+      .writeText(JSON.stringify(meta, null, 2));
+    console.log(
+      `${wr}`,
+      'bundled',
+      modTimeByPath.size,
+      'files in',
+      bundleFileName,
+      'at',
+      bundleTime,
+    );
+  };
+
+  const validate = async targetName => {
+    const metaRd = wr.readOnly().neighbor(`bundle-${targetName}-meta.json`);
```

Comment:
Can we give the user more control over the location of the cache and its metadata? I imagine that’ll be necessary for arrangements where build products either are or aren’t checked in, where CI needs to be able to capture them, &c.

Reply:
I added:

```
  const {
    toBundleName = n => `bundle-${n}.js`,
    toBundleMeta = n => `bundle-${n}-meta.js`,
  } = opts;
```

That doesn't work nicely with putting metadata in different directories, but I'm not a fan of designing based on "I imagine...". :)